alloc_sockaddr: handle abstract paths (v2)

timbertson opened the pull request:
Attempting #2233 again.

I got the tests working on MacOS, but they only failed due to a logic error - the path of an unnamed socket was being returned as a string of 14 null bytes, rather than `""` or raising EAFNOSUPPORT. Since the first byte was null, it looked like a perfectly valid (if weird) abstract socket name. I've had to add an ugly `#ifdef __linux__` now, which makes me sad.

For context, support for returning `""` for an unnamed socket was originally added in https://caml.inria.fr/mantis/view.php?id=7039 (@xavierleroy). The dangerous part of that bug (reading past the end of the allocated space) is already fixed by using `strnlen`, so we _could_ decide that you once again get garbage for an unnamed socket. But it's probably more useful to keep the current behaviour, even if the code is a bit awkward.

/cc @diml @shindere. If you're able to run this through pre-check and find out (a) the backtrace and (b) the test output, that would be super useful. I'm assuming EAFNOSUPPORT is coming from the case where the sender has an unnamed socket. But I'd like to figure out which function is actually raising it.

timbertson committed:
alloc_sockaddr: handle unbound sockets on MacOS

timbertson committed:
Revert "Revert "alloc_sockaddr: handle abstract paths""

This reverts commit 0a3525cc51547a2cc958c9c22a2956f8f611b402.

shindere commented:
Yes I'll do the precheck and report its results here. Thanks!

shindere commented:
OK this is now going through precheck as build #188(https://ci.inria.fr/ocaml/job/precheck/188)

ghost commented:
precheck looks good, there are a few failures but they look unrelated to this PR.

AFAIK, support for abstract socket is a Linux specific feature, so the `#ifdef __linux__` seems reasonable here.

shindere commented:
Jérémie Dimino (2019/02/13 13:35 +0000):
> precheck looks good, there are a few failures but they look unrelated
> to this PR.

The one on zsystems seems rather related!

Investing what exactly happens.

shindere commented:
OK so on zsystems, which is a Linux machine, the test
tests/lib-unix/unix-socket/recvfrom_unix.ml fails in native mode and
here is the output of the program when run:

"ocaml-test-socket-unix" as "ocaml-test-socket-unix": OK
""Fatal error: exception Unix.Unix_error(Unix.EAFNOSUPPORT, "", "")

ghost commented:
Ah indeed, I missed that. @timbertson do you have enough information to understand and fix the issue?

timbertson commented:
Thanks, @shindere. Is it possible to get a backtrace? In particular it could be either the `sendto` or the `recvfrom` call which could be raising. I can't log into https://ci.inria.fr/ocaml/job/precheck/188 so I don't know if there are more details in there which I can't see.

If it's not easy to get a backtrace, I could push a temporary commit with a bit of extra logging between the calls to find out how far it gets...

You mention zsystems, is that https://en.wikipedia.org/wiki/Linux_on_z_Systems ?

timbertson committed:
WIP debugging

timbertson committed:
add try/catch

shindere commented:
Tim Cuthbertson (2019/02/13 13:14 -0800):
> Thanks, @shindere. Is it possible to get a backtrace? In particular it
> could be either the `sendto` or the `recvfrom` call which could be
> raising. I can't log into https://ci.inria.fr/ocaml/job/precheck/188
> so I don't know if there are more details in there which I can't see.

Well it's very odd. FWIW, the test fails on my machine, too.
To try to figure out more precisely what is going on I did an strace:

```
unlink("ocaml-test-socket-unix")        = -1 ENOENT (No such file or directory)
socket(AF_UNIX, SOCK_DGRAM, 0)          = 3
bind(3, {sa_family=AF_UNIX, sun_path="ocaml-test-socket-unix"}, 24) = 0
sendto(3, "t", 1, 0, {sa_family=AF_UNIX, sun_path="ocaml-test-socket-unix"}, 24) = 1
recvfrom(3, "t", 1024, 0, {sa_family=AF_UNIX, sun_path="ocaml-test-socket-unix"}, [112->25]) = 1
write(1, "\"ocaml-test-socket-unix\" as \"oca"..., 57"ocaml-test-socket-unix" as "ocaml-test-socket-unix": OK
) = 57
socket(AF_UNIX, SOCK_DGRAM, 0)          = 4
sendto(4, "t", 1, 0, {sa_family=AF_UNIX, sun_path="ocaml-test-socket-unix"}, 24) = 1
recvfrom(3, "t", 1024, 0, 0x7ffd5f310e40, [112->0]) = 1
write(1, "\"\" as \"Unix.Unix_error(Unix.EAFN"..., 59"" as "Unix.Unix_error(Unix.EAFNOSUPPORT, \"\", \"\")": OK
) = 59
close(4)                                = 0
close(3)                                = 0
unlink("ocaml-test-socket-unix")        = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Does that help?

the other thing I find odd is that it fails only in native mode, while
the bytecode test passes.

> You mention zsystems, is that
> https://en.wikipedia.org/wiki/Linux_on_z_Systems ?

I think so, it's a s390x architecture, but given that the test also
fails on my regular x86_64 I am not sure how relevant that is, finally.

timbertson commented:
Thanks @shindere , I can reproduce now too with the latest two commits. I meant to add this comment when I pushed them, but github was timing out:

---

OK, this is getting very weird. I added some extra logging and a try/catch so I could see exactly where it failed, and it suddenly started failing locally on linux with EAFNOSUPPORT. It _only_ failed with native compilation, the bytecode mode still passed.

I reduced the diff to its minimal form, and it ended up that **if I surround the `recvfrom` line with a try/catch, it throws (and gets caught). If I don't surround it, no error is thrown.**

Debug changes (still passes): tree 718e21e25d248ab86688593fda14cc59f05d83fb
Minimal diff (fails due to EAFNOSUPPORT being thrown and caught): commit 8283b3e8c06f4b18a787ad7332aabcc323ddb66a

Further, I added some debug logging to sockaddr.c itself to print out the `sa_family` value in alloc_sockaddr (commented out in the above commits, because it breaks the test output checks). With the code that passes, it prints `address family: 1` for both test cases.

With the try/catch, it prints `1` for the first test case (a named socket), then `0` for the second test case (the unnamed socket). This causes that function to raise EAFNOSUPPORT, since `0` isn't a known address family and it rightly falls through to the `default` switch case.

So my question is: how on earth could adding a try/catch affect that? This seems like either I'm forgetting something dumb, or it's quite a nasty bug.

I cherry-picked my branch onto 4.07.1 to see if it was a recent bug, but it has the same behaviour there.

----

Concrete steps for anyone who wants to reproduce (please reproduce, tell me I'm not going mad :wink: ):

```
$ git checkout 8283b3e8c06f4b18a787ad7332aabcc323ddb66a^
$ make world.opt
$ ocamltest/ocamltest testsuite/tests/lib-unix/unix-socket/recvfrom_unix.ml -e
# (passes)

$ git checkout 8283b3e8c06f4b18a787ad7332aabcc323ddb66a
$ ocamltest/ocamltest testsuite/tests/lib-unix/unix-socket/recvfrom_unix.ml -e
# (fails)
```

shindere commented:
Well if you are going crazy then there are two of us I'd say. ;-)

I'll talk with Xavier and Damien tomorrow and see whether they have
insights.

shindere commented:
Talked with Xavier as announced. the fact that a try with could modify
the program's behaviour was a hint for Xavier, he immediately suggested
that the program was somehow trying to read data from the stack that had
not been previously initialized (since the try with construct introduces
shifts on the flag it explains the differences in behaviour.)

After having spent some time unsuccessfully trying to use clang's
sanitizers, I thought about valgrind. I recompiled the whole compiler
with flags -g -O0, used ocamltest to get get the test program compiled
once and then ran the compiled test program under valgrind. Here is what
I get:

```
==13910== Memcheck, a memory error detector
==13910== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==13910== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==13910== Command: ./recvfrom_unix.opt
==13910== 
"ocaml-test-socket-unix" as "ocaml-test-socket-unix": OK
==13910== Conditional jump or move depends on uninitialised value(s)
==13910==    at 0x19FDFD: alloc_sockaddr (socketaddr.c:111)
==13910==    by 0x19F390: unix_recvfrom (sendrecv.c:73)
...
```

So my understanding is that, for some reason yet to be understood,
`unix_recvfrom` does not provide enough bytes for `adr->s_gen.sa_family`
to be filled with something meaningful when arriving at the switch.

Does this help at least a little bit?

shindere commented:
By the way, just for the record I ran valgrind as follows:

```
valgrind --track-origins=yes ./recvfrom_unix.opt
```

timbertson commented:
That's incredibly helpful, thanks! I've got a few ideas now, I'll do some digging and report back.

timbertson committed:
recvfrom: protect against uninitialized sa_family

timbertson committed:
cleanup

timbertson commented:
OK, I think I've figured it out: the existing code only ever worked by chance. When the sender is unbound, `recvfrom` sets adr_len to `0`, and doesn't write anything to the address pointer. We then switch on the first field of that address (`sa_family`), which is uninitialized.

Since this can _only_ happen for `AF_UNIX`, I've added a guard so that if `adr_len` is zero, we set sa_family explicitly to `AF_UNIX` (since the rest of the code already handles a zero-length address correctly).

It might be more correct to use `getsockname` to get the receiver's `sa_family` and use that in `unix_recvfrom`, but that would have to be done differently at each callsite to `alloc_sockaddr`.

@shindere could you please run the latest commit through pre-check? If everything passes I'll squash the commits and we'll be good to go :grinning:

shindere commented:
Tim Cuthbertson (2019/02/15 16:21 -0800):
> OK, I think I've figured it out: the existing code only ever worked by
> chance.

Yes, that is my understanding, too.

> When the sender is unbound, `recvfrom` sets adr_len to `0`, and
> doesn't write anything to the address pointer.
> We then switch on the first field of that address (`sa_family`),
> which is uninitialized.

Right.

> Since this can _only_ happen for `AF_UNIX`, I've added a guard so that
> if `adr_len` is zero, we set sa_family explicitly to `AF_UNIX` (since
> the rest of the code already handles a zero-length address correctly).

Well, I understand why and how this works, practically, but in principle
I am wondering whether it wouldn't be possible to take into account that
the length is zero and nothing has been returned to not even do the
switch, rather than "pretending" some data has been returned?

timbertson commented:
The trouble is you still need to return a Unix address with an empty path
(since this function isn't allowed to return None). Repeating the code to
build one of those in the branch when adr_len is zero would work, but it's
extra (repeated) code to achieve the same effect.

The existing code is careful when bounds checking the length of the
returned path information, so it seemed simpler to keep all that logic in
one place.

>

shindere commented:
Tim Cuthbertson (2019/02/18 11:29 +0000):
> The trouble is you still need to return a Unix address with an empty path
> (since this function isn't allowed to return None). Repeating the code to
> build one of those in the branch when adr_len is zero would work, but it's
> extra (repeated) code to achieve the same effect.
> 
> The existing code is careful when bounds checking the length of the
> returned path information, so it seemed simpler to keep all that logic in
> one place.

Well, if it makes sense, then I'd suggest to put this logic into a
separate funciton and ot use it in both places but if that does not seem
doable then I'm fine with your solution.

timbertson committed:
extract alloc_unix_sockaddr function

timbertson commented:
That seems reasonable, like this? (d75fc80)

I haven't worked with ocaml's C bindings much, so I followed the existing style and hopefully I'm not doing anything silly ;)

ghost commented:
Yes, that looks correct. A more idiomatic version would be:

```c
value alloc_unix_sockaddr(value path) {
￼  CAMLparam1(path);
  CAMLlocal1(res);
  res = caml_alloc_small(1, 0);
  Field(res,0) = path;
  CAMLreturn(res);
}
```

The Begin/End_roots macros are more low-level, you need to understand more in detail what they are doing to check the correctness of the code.

I restarted precheck with the tip of this PR.

timbertson committed:
make alloc_unix_sockaddr more idiomatic

hannesm commented on otherlibs/unix/socketaddr.c:
```diff
@@ -103,12 +103,11 @@ void get_sockaddr(value mladr,
-    Field(res,0) = path;
-  End_roots();
-  return res;
+￼ CAMLparam1(path);
```
seems like a [unicode object replacement character](https://www.fileformat.info/info/unicode/char/fffc/index.htm) slipped into the previous line (first codepoint), should instead be a normal whitespace (0x20).

timbertson committed:
fix indentation

ghost commented:
The tip of this PR looks ready to me. @shindere does it look good to you as well?

shindere commented:
Jérémie Dimino (2019/02/21 05:24 -0800):
> The tip of this PR looks ready to me. @shindere does it look good to
> you as well?

Yes, sorry for the lag! I think this has already been approved so feel
free to merge but perhaps make sure to squash all the commits.

ghost commented:
Sounds good!

ghost commented:
BTW, I moved the changelog entry to trunk since this didn't make it to 4.08.

shindere commented:
Jérémie Dimino (2019/02/21 07:40 -0800):
> BTW, I moved the changelog entry to trunk since this didn't make it to
> 4.08.

OK thanks!

I think we do not want it in 4.08, do we?

timbertson-zd commented:
Thanks for your help @shindere and @diml! :)

> I think we do not want it in 4.08, do we?

I don't think so, it's just that the initial PR snuck into 4.08 so I put the Changes entry there the first time around.
